fastestimator/op/tensorop/model/update.py

```diff
@@ -74,22 +107,130 @@ def fe_retain_graph(self, retain: Optional[bool] = None) -> Optional[bool]:
+
+        elif self.framework == "torch":
+            trainable_params = [p for p in self.model.parameters() if p.requires_grad]
+            try:
+                gradients = get_gradient(loss, trainable_params, retain_graph=self.retain_graph)
+            except RuntimeError as err:
+                if err.args and isinstance(err.args[0], str) and err.args[0].startswith(
+                        'one of the variables needed for gradient computation has been modified by an inplace operation'
+                ):
+                    raise RuntimeError(
+                        "When computing gradients for '{}', some variables it relied on during the forward pass had"
+                        " been updated. Consider setting defer=True in earlier UpdateOps related to models which "
+                        "interact with this one.".format(self.model.model_name))
+                raise err
+
+        else:
+            raise ValueError(f"Unrecognized framework {self.framework}")
+
+        return gradients
+
+    def _gradient_postprocess(self, gradients, state):
+        """ Gradient postprocess for multi-GPU and mixed-precision training
+        """
+        if self.framework == "tf":
+            if self.gradients is not None:  # user provide gradients
+                strategy = tf.distribute.get_strategy()  # need tape.stop_recording() ?
+                # for multi-gpu training, the gradient will be combined by sum, normalize the gradient
+                if isinstance(strategy, tf.distribute.MirroredStrategy):
+                    gradients = gradients / strategy.num_replicas_in_sync
+
+            # scale down gradient to balance scale-up loss
+            if self.model.mixed_precision:
+                with state["tape"].stop_recording():
+                    gradients = self.model.current_optimizer.get_unscaled_gradients(gradients)
+
+        elif self.framework == "torch":
+            pass
+
+        else:
+            raise ValueError(f"Unrecognized framework {self.framework}")
+
+        return gradients
+
+    def _merge_grad_update(self, gradients, state):
+        current_grad = gradients
+        if self.framework == "tf":
+            self.step.assign_add(1)
+            # add current gradient to the accumulated gradient
+            for gs, g in zip(self.grad_sum, current_grad):
+                gs.assign_add(g)
+            if self.step % self.merge_grad == 0:
+                average_grad = [gs / self.merge_grad for gs in self.grad_sum]
+                # apply update with cumulative gradient
+                update_model(model=self.model,
+                             gradients=average_grad,
+                             tape=state["tape"],
                              defer=self.defer,
                              deferred=state["deferred"])
+
+                # zero-out cumulative gradient
+                for gs in self.grad_sum:
+                    gs.assign_sub(gs)
+
+        elif self.framework == "torch":
```

Comment:
the control flow here between tf and torch looks basically identical. can we get it into a single branch for better maintainability?

Reply:
fixed